{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Global scRNAseq and functional data curation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# main output file with all samples merged\n",
    "merged_adata_file_name = \"all_merged.h5ad\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import r\n",
    "from rpy2.robjects.packages import importr\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "import hdf5plugin\n",
    "import path_config\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define paths to original data.<br>\n",
    "Original data are .rds files with results of zUMIs. umicount exon all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "received_data1_path = os.path.join(path_config.data_path,\"SS3_22_291.dgecounts.rds\")\n",
    "received_data2_path = os.path.join(path_config.data_path,\"SS3_23_049.dgecounts.rds\") \n",
    "received_data3_path = os.path.join(path_config.data_path,\"SS3_23_127.dgecounts.rds\") \n",
    "received_data4_path = os.path.join(path_config.data_path,\"SS3_23_193.dgecounts.rds\")\n",
    "received_data5_path = os.path.join(path_config.data_path,\"SS3_23_195.dgecounts.rds\")\n",
    "received_data6_path = os.path.join(path_config.data_path,\"SS3_23_325.dgecounts.rds\")\n",
    "received_data7_path = os.path.join(path_config.data_path,\"SS3_23_327.dgecounts.rds\")\n",
    "received_data8_path = os.path.join(path_config.data_path,\"P30855_101.dgecounts.rds\")\n",
    "received_data9_path = os.path.join(path_config.data_path,\"P30855_103.dgecounts.rds\")\n",
    "received_data10_path = os.path.join(path_config.data_path,\"P30855_105.dgecounts.rds\")\n",
    "received_data11_path = os.path.join(path_config.data_path,\"P31406_101.dgecounts.rds\")\n",
    "\n",
    "# create a list of all original paths to automate reading\n",
    "received_data_paths_list = [received_data1_path,received_data2_path,\n",
    "                      received_data3_path,received_data4_path,\n",
    "                      received_data5_path,received_data6_path,\n",
    "                      received_data7_path,received_data8_path,\n",
    "                      received_data9_path,received_data10_path,\n",
    "                      received_data11_path\n",
    "                      ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define paths to barcodes for each sample.<br>\n",
    "These are the files from scilifeLab<br>\n",
    "bc_set + tab + well_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# txt file with well barcodes. the text contains the cell barcode corresponding to each position in the plate\n",
    "well_barcode_sample1 = os.path.join(path_config.data_path, \"SS3_22_291.well_barcodes.txt\")\n",
    "well_barcode_sample2 = os.path.join(path_config.data_path, \"SS3_23_049.well_barcodes.txt\")\n",
    "well_barcode_sample3 = os.path.join(path_config.data_path, \"SS3_23_127.well_barcodes.txt\")\n",
    "well_barcode_sample4 = os.path.join(path_config.data_path, \"SS3_23_193.well_barcodes.txt\")\n",
    "well_barcode_sample5 = os.path.join(path_config.data_path, \"SS3_23_195.well_barcodes.txt\")\n",
    "well_barcode_sample6 = os.path.join(path_config.data_path, \"SS3_23_325.well_barcodes.txt\")\n",
    "well_barcode_sample7 = os.path.join(path_config.data_path, \"SS3_23_327.well_barcodes.txt\")\n",
    "well_barcode_sample8 = os.path.join(path_config.data_path, \"P30855_101.well_barcodes.txt\")\n",
    "well_barcode_sample9 = os.path.join(path_config.data_path, \"P30855_103.well_barcodes.txt\")\n",
    "well_barcode_sample10 = os.path.join(path_config.data_path, \"P30855_105.well_barcodes.txt\")\n",
    "well_barcode_sample11 = os.path.join(path_config.data_path, \"P31406_101.well_barcodes.txt\")\n",
    "\n",
    "# create a list of barcodes per sample\n",
    "all_well_barcodes_per_sample = [well_barcode_sample1,well_barcode_sample2,\n",
    "                                well_barcode_sample3,well_barcode_sample4,\n",
    "                                well_barcode_sample5,well_barcode_sample6,\n",
    "                                well_barcode_sample7,well_barcode_sample8,\n",
    "                                well_barcode_sample9,well_barcode_sample10,\n",
    "                                well_barcode_sample11\n",
    "                                ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define paths to files with functional experimental data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# upload excel with red and green flourecence from campari (has information about position in plate and fluorecence level)\n",
    "florecence_1_file_path = os.path.join(path_config.data_path, \"well_1_data_base_labeled.csv\")\n",
    "florecence_2_file_path = os.path.join(path_config.data_path, \"well_2_data_base_labeled.csv\")\n",
    "florecence_3_file_path = os.path.join(path_config.data_path, \"well_3_data_base_labeled.csv\")\n",
    "florecence_4_file_path = os.path.join(path_config.data_path, \"well_4_data_base_labeled.csv\")\n",
    "florecence_5_file_path = os.path.join(path_config.data_path, \"well_5_data_base_labeled.csv\")\n",
    "#######################################################################################################################################\n",
    "###BE CAREFULL WITH PROJECT NAME AND PLATE NUMBER ASSIGNMENT!\n",
    "### WE SWAP WERE WHAT IS 6 AND 7 IN THE STIMULI FILE SO BARCODES ARE WITH THEIR CORRECT STIMULI ID\n",
    "#######################################################################################################################################\n",
    "florecence_6_file_path = os.path.join(path_config.data_path, \"well_7_data_base_labeled.csv\") # swapped\n",
    "florecence_7_file_path = os.path.join(path_config.data_path, \"well_6_data_base_labeled.csv\") # swapped\n",
    "florecence_8_file_path = os.path.join(path_config.data_path, \"well_8_data_base_labeled.csv\")\n",
    "florecence_9_file_path = os.path.join(path_config.data_path, \"well_9_data_base_labeled.csv\")\n",
    "florecence_10_file_path = os.path.join(path_config.data_path, \"well_10_data_base_labeled.csv\")\n",
    "florecence_11_file_path = os.path.join(path_config.data_path, \"well_11_data_base_labeled.csv\")\n",
    "\n",
    "# create a list of files with metadata from the lab\n",
    "all_fluorescence_paths = [florecence_1_file_path,florecence_2_file_path,\n",
    "                        florecence_3_file_path,florecence_4_file_path,\n",
    "                        florecence_5_file_path,florecence_6_file_path,\n",
    "                        florecence_7_file_path,florecence_8_file_path,\n",
    "                        florecence_9_file_path,florecence_10_file_path,\n",
    "                        florecence_11_file_path\n",
    "                        ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read and organize UMI counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing sample 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: Matrix\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes, cells\n",
      "(27978, 384)\n",
      "\n",
      "Processing sample 2\n",
      "genes, cells\n",
      "(29575, 384)\n",
      "\n",
      "Processing sample 3\n",
      "genes, cells\n",
      "(28424, 384)\n",
      "\n",
      "Processing sample 4\n",
      "genes, cells\n",
      "(29285, 384)\n",
      "\n",
      "Processing sample 5\n",
      "genes, cells\n",
      "(28390, 384)\n",
      "\n",
      "Processing sample 6\n",
      "genes, cells\n",
      "(28190, 384)\n",
      "\n",
      "Processing sample 7\n",
      "genes, cells\n",
      "(28706, 384)\n",
      "\n",
      "Processing sample 8\n",
      "genes, cells\n",
      "(29363, 384)\n",
      "\n",
      "Processing sample 9\n",
      "genes, cells\n",
      "(29682, 384)\n",
      "\n",
      "Processing sample 10\n",
      "genes, cells\n",
      "(28587, 384)\n",
      "\n",
      "Processing sample 11\n",
      "genes, cells\n",
      "(28914, 384)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Create a list of all data frames. One data frame per sample\n",
    "all_samples_dfs = []\n",
    "# check how many cells were in each of the data sets\n",
    "all_total_barcodes = []\n",
    "for i,raw_R_data in enumerate(received_data_paths_list):\n",
    "    print(f\"Processing sample\",i+1)\n",
    "    #print(received_data_paths_list[i])\n",
    "    # Read the RDS file\n",
    "    r_readRDS = robjects.r['readRDS']\n",
    "    received_data = r_readRDS(raw_R_data)\n",
    "    # Extract \"umicount\" data\n",
    "    new_data = received_data.rx2(\"umicount\")\n",
    "    exon = new_data.rx2(\"exon\").rx2(\"all\")\n",
    "    # Convert dgCMatrix to a dense matrix in R\n",
    "    as_matrix = robjects.r['as.matrix']\n",
    "    dense_matrix = as_matrix(exon)  # Convert to dense matrix\n",
    "    # Convert dense matrix to NumPy array\n",
    "    exon_matrix = np.array(dense_matrix)\n",
    "    \n",
    "    # Extract row and column names\n",
    "    base = importr('base')\n",
    "    rownames = list(base.rownames(exon))\n",
    "    colnames = list(base.colnames(exon))\n",
    "\n",
    "    # Create a pandas DataFrame\n",
    "    my_df = pd.DataFrame(exon_matrix, index=rownames, columns=colnames)\n",
    "\n",
    "    # check how many cells were in each of the data sets\n",
    "    total_barcodes = my_df.shape[1]\n",
    "    all_total_barcodes.append(total_barcodes)\n",
    "    # they are different sizes\n",
    "    all_samples_dfs.append(my_df)\n",
    "    print(\"genes, cells\")\n",
    "    print(my_df.shape)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AACCATCGGCAACTACCACT</th>\n",
       "      <th>AACCATCGGCAAGTTATCGG</th>\n",
       "      <th>AACCATCGGCCACAATCCAC</th>\n",
       "      <th>AACCATCGGCCACTAACCGG</th>\n",
       "      <th>AACCATCGGCCCGTTCGTAT</th>\n",
       "      <th>AACCATCGGCCGACGGTTGT</th>\n",
       "      <th>AACCATCGGCCGAGTCATTA</th>\n",
       "      <th>AACCATCGGCGAAGCTAGCC</th>\n",
       "      <th>AACCATCGGCGTAACGTCTA</th>\n",
       "      <th>AACCATCGGCTCCAATGCAA</th>\n",
       "      <th>...</th>\n",
       "      <th>TTGTCTGCAAGCACAGTATT</th>\n",
       "      <th>TTGTCTGCAATACAACTAGC</th>\n",
       "      <th>TTGTCTGCAATAGGTGGCGA</th>\n",
       "      <th>TTGTCTGCAATAGTTAGGCT</th>\n",
       "      <th>TTGTCTGCAATCTGCAACAA</th>\n",
       "      <th>TTGTCTGCAATGCATCATAC</th>\n",
       "      <th>TTGTCTGCAATGGCCTTAAT</th>\n",
       "      <th>TTGTCTGCAATTACAACACG</th>\n",
       "      <th>TTGTCTGCAATTCAGGAGAT</th>\n",
       "      <th>TTGTCTGCAATTCTCTGATC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610010K14Rik</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610025J13Rik</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trav6-7-dv9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Usp9y</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vmn1r80</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wfdc11</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zfp36l3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35546 rows × 4224 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               AACCATCGGCAACTACCACT  AACCATCGGCAAGTTATCGG  \\\n",
       "0610005C13Rik                   0.0                   0.0   \n",
       "0610009E02Rik                   0.0                   0.0   \n",
       "0610009L18Rik                   3.0                   0.0   \n",
       "0610010K14Rik                   2.0                   2.0   \n",
       "0610025J13Rik                   0.0                   0.0   \n",
       "...                             ...                   ...   \n",
       "Trav6-7-dv9                     0.0                   0.0   \n",
       "Usp9y                           0.0                   0.0   \n",
       "Vmn1r80                         0.0                   0.0   \n",
       "Wfdc11                          0.0                   0.0   \n",
       "Zfp36l3                         0.0                   0.0   \n",
       "\n",
       "               AACCATCGGCCACAATCCAC  AACCATCGGCCACTAACCGG  \\\n",
       "0610005C13Rik                   0.0                   0.0   \n",
       "0610009E02Rik                   0.0                   0.0   \n",
       "0610009L18Rik                   2.0                   0.0   \n",
       "0610010K14Rik                   4.0                   0.0   \n",
       "0610025J13Rik                   0.0                   0.0   \n",
       "...                             ...                   ...   \n",
       "Trav6-7-dv9                     0.0                   0.0   \n",
       "Usp9y                           0.0                   0.0   \n",
       "Vmn1r80                         0.0                   0.0   \n",
       "Wfdc11                          0.0                   0.0   \n",
       "Zfp36l3                         0.0                   0.0   \n",
       "\n",
       "               AACCATCGGCCCGTTCGTAT  AACCATCGGCCGACGGTTGT  \\\n",
       "0610005C13Rik                   0.0                   0.0   \n",
       "0610009E02Rik                   0.0                   0.0   \n",
       "0610009L18Rik                   1.0                   2.0   \n",
       "0610010K14Rik                   4.0                   3.0   \n",
       "0610025J13Rik                   0.0                   0.0   \n",
       "...                             ...                   ...   \n",
       "Trav6-7-dv9                     0.0                   0.0   \n",
       "Usp9y                           0.0                   0.0   \n",
       "Vmn1r80                         0.0                   0.0   \n",
       "Wfdc11                          0.0                   0.0   \n",
       "Zfp36l3                         0.0                   0.0   \n",
       "\n",
       "               AACCATCGGCCGAGTCATTA  AACCATCGGCGAAGCTAGCC  \\\n",
       "0610005C13Rik                   0.0                   0.0   \n",
       "0610009E02Rik                   0.0                   1.0   \n",
       "0610009L18Rik                   0.0                   0.0   \n",
       "0610010K14Rik                   1.0                   0.0   \n",
       "0610025J13Rik                   0.0                   0.0   \n",
       "...                             ...                   ...   \n",
       "Trav6-7-dv9                     0.0                   0.0   \n",
       "Usp9y                           0.0                   0.0   \n",
       "Vmn1r80                         0.0                   0.0   \n",
       "Wfdc11                          0.0                   0.0   \n",
       "Zfp36l3                         0.0                   0.0   \n",
       "\n",
       "               AACCATCGGCGTAACGTCTA  AACCATCGGCTCCAATGCAA  ...  \\\n",
       "0610005C13Rik                   0.0                   0.0  ...   \n",
       "0610009E02Rik                   0.0                   0.0  ...   \n",
       "0610009L18Rik                   0.0                   0.0  ...   \n",
       "0610010K14Rik                   3.0                   3.0  ...   \n",
       "0610025J13Rik                   0.0                   0.0  ...   \n",
       "...                             ...                   ...  ...   \n",
       "Trav6-7-dv9                     0.0                   0.0  ...   \n",
       "Usp9y                           0.0                   0.0  ...   \n",
       "Vmn1r80                         0.0                   0.0  ...   \n",
       "Wfdc11                          0.0                   0.0  ...   \n",
       "Zfp36l3                         0.0                   0.0  ...   \n",
       "\n",
       "               TTGTCTGCAAGCACAGTATT  TTGTCTGCAATACAACTAGC  \\\n",
       "0610005C13Rik                   0.0                   0.0   \n",
       "0610009E02Rik                   0.0                   0.0   \n",
       "0610009L18Rik                   0.0                   0.0   \n",
       "0610010K14Rik                   2.0                   4.0   \n",
       "0610025J13Rik                   0.0                   0.0   \n",
       "...                             ...                   ...   \n",
       "Trav6-7-dv9                     0.0                   0.0   \n",
       "Usp9y                           0.0                   0.0   \n",
       "Vmn1r80                         0.0                   0.0   \n",
       "Wfdc11                          0.0                   0.0   \n",
       "Zfp36l3                         0.0                   0.0   \n",
       "\n",
       "               TTGTCTGCAATAGGTGGCGA  TTGTCTGCAATAGTTAGGCT  \\\n",
       "0610005C13Rik                   0.0                   0.0   \n",
       "0610009E02Rik                   0.0                   0.0   \n",
       "0610009L18Rik                   3.0                   0.0   \n",
       "0610010K14Rik                   2.0                   8.0   \n",
       "0610025J13Rik                   0.0                   0.0   \n",
       "...                             ...                   ...   \n",
       "Trav6-7-dv9                     0.0                   0.0   \n",
       "Usp9y                           0.0                   0.0   \n",
       "Vmn1r80                         0.0                   0.0   \n",
       "Wfdc11                          0.0                   0.0   \n",
       "Zfp36l3                         0.0                   0.0   \n",
       "\n",
       "               TTGTCTGCAATCTGCAACAA  TTGTCTGCAATGCATCATAC  \\\n",
       "0610005C13Rik                   0.0                   0.0   \n",
       "0610009E02Rik                   0.0                   0.0   \n",
       "0610009L18Rik                   1.0                   0.0   \n",
       "0610010K14Rik                  10.0                   5.0   \n",
       "0610025J13Rik                   0.0                   0.0   \n",
       "...                             ...                   ...   \n",
       "Trav6-7-dv9                     0.0                   0.0   \n",
       "Usp9y                           0.0                   0.0   \n",
       "Vmn1r80                         0.0                   0.0   \n",
       "Wfdc11                          0.0                   0.0   \n",
       "Zfp36l3                         0.0                   0.0   \n",
       "\n",
       "               TTGTCTGCAATGGCCTTAAT  TTGTCTGCAATTACAACACG  \\\n",
       "0610005C13Rik                   0.0                   0.0   \n",
       "0610009E02Rik                   0.0                   0.0   \n",
       "0610009L18Rik                   2.0                   0.0   \n",
       "0610010K14Rik                   1.0                   6.0   \n",
       "0610025J13Rik                   0.0                   0.0   \n",
       "...                             ...                   ...   \n",
       "Trav6-7-dv9                     0.0                   0.0   \n",
       "Usp9y                           0.0                   0.0   \n",
       "Vmn1r80                         0.0                   0.0   \n",
       "Wfdc11                          0.0                   0.0   \n",
       "Zfp36l3                         0.0                   0.0   \n",
       "\n",
       "               TTGTCTGCAATTCAGGAGAT  TTGTCTGCAATTCTCTGATC  \n",
       "0610005C13Rik                   0.0                   0.0  \n",
       "0610009E02Rik                   0.0                   1.0  \n",
       "0610009L18Rik                   0.0                   1.0  \n",
       "0610010K14Rik                   5.0                   3.0  \n",
       "0610025J13Rik                   0.0                   1.0  \n",
       "...                             ...                   ...  \n",
       "Trav6-7-dv9                     0.0                   0.0  \n",
       "Usp9y                           0.0                   0.0  \n",
       "Vmn1r80                         0.0                   0.0  \n",
       "Wfdc11                          0.0                   0.0  \n",
       "Zfp36l3                         0.0                   0.0  \n",
       "\n",
       "[35546 rows x 4224 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge whole dataframe, so that all will have the same amount of genes. If some did not contain certain gene, count=0\n",
    "unified_df = pd.concat(all_samples_dfs,axis=1)\n",
    "# replace all NaN values with zero counts\n",
    "unified_df = unified_df.fillna(0.0)\n",
    "unified_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# decompose from unified\n",
    "all_decomposed_samples = []\n",
    "for i, total_barcodes in enumerate(all_total_barcodes):\n",
    "    if i == 0: # for the first one\n",
    "        sample_df = unified_df.iloc[:,0:all_total_barcodes[i]]\n",
    "        till_col = all_total_barcodes[i]+all_total_barcodes[i+1]\n",
    "    elif i == 1:\n",
    "        sample_df = unified_df.iloc[:,all_total_barcodes[0]:till_col]\n",
    "        from_col = till_col\n",
    "        till_col = from_col+all_total_barcodes[i+1]\n",
    "    elif i == len(all_total_barcodes)-1:\n",
    "        sample_df = unified_df.iloc[:,from_col:till_col]\n",
    "    else:\n",
    "        sample_df = unified_df.iloc[:,from_col:till_col]\n",
    "        from_col = till_col\n",
    "        till_col = from_col+all_total_barcodes[i+1]\n",
    "    #print(sample_df.shape[1]) # number of cells\n",
    "    all_decomposed_samples.append(sample_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create single adata objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert all df to AnnData objects \n",
    "# export to h5ad files\n",
    "all_adatas = []\n",
    "for i, df in enumerate(all_decomposed_samples):\n",
    "    # transpose before creating anndata object, so that gene ids are columns and barcodes are rows\n",
    "    transposed_rds_sample= all_decomposed_samples[i].transpose()# Create an AnnData object\n",
    "    adata = ad.AnnData(\n",
    "        X=transposed_rds_sample.values,  # Your data as numpy array\n",
    "        obs={'barcode': transposed_rds_sample.index},\n",
    "        var={'gene_id': transposed_rds_sample.columns}  \n",
    "    )\n",
    "    # add sample metadata\n",
    "    sample = \"Sample_\" + str(i+1)\n",
    "    adata.obs['sample'] = sample\n",
    "\n",
    "    all_adatas.append(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge functional experimental data with scRNAseq data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Helper function for adding functional tags to scRNAseq data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# returns ordered well ids, stimulus and fluorescence from each sample\n",
    "def barcode_check(barcode_list,mergedRes):\n",
    "    \n",
    "    well_id = [\"NA\" for el in barcode_list] # these are strings (i.e \"A1\")\n",
    "    red_f = np.zeros((len(barcode_list),))\n",
    "    green_f = np.zeros((len(barcode_list),))\n",
    "    stimulus = np.zeros((len(barcode_list),))\n",
    "    plate_number = np.zeros((len(barcode_list),))\n",
    "    area = np.zeros((len(barcode_list),))\n",
    "    cell_id = np.zeros((len(barcode_list),))\n",
    "\n",
    "\n",
    "    for i in range(len(barcode_list)):\n",
    "        x = barcode_list[i]\n",
    "        index = barcode_list.index(x)\n",
    "        if len(np.where(np.isnan(mergedRes['stimulus'].where(mergedRes['bc_set'] == x))==False)):\n",
    "            if np.where(mergedRes['bc_set'] == x)[0]:\n",
    "                position = np.where(np.isnan(mergedRes['stimulus'].where(mergedRes['bc_set'] == x))==False)[0][0]\n",
    "                stimulus[index] = mergedRes['stimulus'].where(mergedRes['bc_set'] == x)[position]\n",
    "                red_f[index] = mergedRes['red_f'].where(mergedRes['bc_set'] == x)[position]\n",
    "                green_f[index] = mergedRes['green_f'].where(mergedRes['bc_set'] == x)[position]\n",
    "                well_id[index] = mergedRes['well_id'].where(mergedRes['bc_set'] == x)[position]\n",
    "                plate_number[index] = mergedRes['well_number'].where(mergedRes['bc_set'] == x)[position]\n",
    "                area[index] = mergedRes['AREA'].where(mergedRes['bc_set'] == x)[position]\n",
    "                cell_id[index] = mergedRes['cell_number'].where(mergedRes['bc_set'] == x)[position]\n",
    "                \n",
    "    # replace all nans to zeros\n",
    "    np.nan_to_num(well_id,copy=False)\n",
    "    np.nan_to_num(stimulus,copy=False)\n",
    "    np.nan_to_num(red_f,copy=False)\n",
    "    np.nan_to_num(green_f,copy=False)\n",
    "    np.nan_to_num(plate_number,copy=False)\n",
    "    np.nan_to_num(area,copy=False)\n",
    "    np.nan_to_num(cell_id,copy=False)\n",
    "\n",
    "    return well_id,stimulus,red_f,green_f,plate_number,area, cell_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " \n",
    "# Merge all into one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect all infomration per sample into single adata objects\n",
    "all_complete_adatas = []\n",
    "# in the end we will have merged_adata\n",
    "merged_adata = None\n",
    "for i, sample in enumerate(all_adatas):\n",
    "    current_adata = all_adatas[i]\n",
    "    flouresence_sample = pd.read_csv(all_fluorescence_paths[i])\n",
    "    well_df_sample = pd.read_csv(all_well_barcodes_per_sample[i], delimiter = \"\\t\")\n",
    "    # read barcodes in the same order as in the anndata object\n",
    "    barcodes = [el[0] for el in list(current_adata.obs.values)]\n",
    "    # merge data base. Merge information about cell barcode and fluorecence \n",
    "    #(using as common demonimator the position in the plate)\n",
    "    mergedRes = pd.merge(well_df_sample , flouresence_sample, on ='well_id')\n",
    "    well_id,stimulus,red_f,green_f,plate_number,area, cell_id = barcode_check(barcodes,mergedRes)\n",
    "    current_adata.var['gene_name'] = current_adata.var['gene_id']\n",
    "    current_adata.var.index = current_adata.var[\"gene_name\"]\n",
    "    ### add campari information to adata structure\n",
    "    current_adata.obs['stimulus'] = stimulus\n",
    "    current_adata.obs['red'] = red_f\n",
    "    current_adata.obs['green'] = green_f\n",
    "    current_adata.obs['well_id'] = well_id\n",
    "    current_adata.obs['plate_number'] = plate_number\n",
    "    current_adata.obs['area'] = area\n",
    "    current_adata.obs['cell_id'] = cell_id\n",
    "\n",
    "    # add to list of all adata objects\n",
    "    all_complete_adatas.append(current_adata)\n",
    "\n",
    "    # start concatenating when at least 2 adata objects were in all_complete_adatas\n",
    "    if i > 0 and i < (len(all_adatas)):\n",
    "        if i == 1:\n",
    "            sample_1 = \"sample1\"\n",
    "            sample_2 = \"sample2\"\n",
    "            merged_adata = all_complete_adatas[0].concatenate(all_complete_adatas[i], join='outer', batch_categories=[sample_1, sample_2], index_unique='-')\n",
    "        else:\n",
    "            sample_1 = sample_1+str(i)\n",
    "            sample_2 = \"sample\"+str(i+1)\n",
    "            merged_adata = merged_adata.concatenate(all_complete_adatas[i], join='outer', batch_categories=[sample_1, sample_2], index_unique='-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cells: 4224\n",
      "Number of genes: 35546\n"
     ]
    }
   ],
   "source": [
    "# assign our merged data\n",
    "adata = merged_adata.copy()\n",
    "\n",
    "adata.var_names_make_unique()  # this is unnecessary if using `var_names='gene_ids'` in `sc.read_10x_mtx`\n",
    "print('Number of cells:', adata.X.shape[0]) # len(adata.obs_names)\n",
    "print('Number of genes:', adata.X.shape[1]) # len(adata.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4224 × 35546\n",
       "    obs: 'barcode', 'sample', 'stimulus', 'red', 'green', 'well_id', 'plate_number', 'area', 'cell_id', 'batch'\n",
       "    var: 'gene_id', 'gene_name'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export merged anndata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export merged data\n",
    "adata.write_h5ad(\n",
    "    os.path.join(path_config.output_path,merged_adata_file_name)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
